Added support for emoticons

moe-25 opened the pull request:
This adds support for emoticons resp. characters above 0xFFFF in general. It might be partially related to issue #145. Since PDF cannot handle Unicode surrogates directly, they have to be re-mapped to fit 16-bit. For simplicity, each character is mapped the first time it is used, allowing PDF documents with 655535 __different__ characters per font. Even if glyphs are not present in the font, they can be made visible by copy-and-paste. These changes could probably break other things (including various tests).

* `text()` and `cell()` must remap all characters in incoming strings
* `TTFontFile.makeSubset()` must be informed where to place each glyph in the embedded font
* `TTFontFile` must consider CMAP12 tables to actually reach these glyphs
* Remapping must be taken into account when creating the charWidth definition
* Instead of _bfrange_, _bfchar_ is used to support PDF readers when searching/copying.

There would still be a need for optimization, such as replacing the `subset` array with a dictionary, shrinking the subset when `alias_nb_pages` is used and so on. How do you think about this approach?

Demo:
```
x = fpdf.FPDF()
x.add_font('DejaVuSans', '', 'DejaVuSans.ttf', uni=True)
x.set_font('DejaVuSans')
x.add_page()

x.cell(0, None, "Page {nb}", align="C", ln=1)

x.cell(50, 50, "That's copyable \U0001F5B1")
x.text(100, 50, "That's cool \U0001F60E")

x.output('Test.pdf')
```

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -1066,7 +1066,14 @@ def add_font(self, family, style="", fname=None, uni=False):
+            if self.str_alias_nb_pages:
+                subset_max = 57
+                for char in self.str_alias_nb_pages:
+                    subset_max = max(subset_max, ord(char))
```
Why this change?
In practice, `sbarr` wille become `list(range(125))` if `str_alias_nb_pages` is set.

I still wonder where this value of `57` is coming from...

Also, it terms of code, this would be shorter:
```python
subset_max = max(map(ord, self.str_alias_nb_pages))
```

moe-25 replied:
Substituting page numbers became more complicated. When `_substitute_page_number()` calls `str.replace()`, it is no longer known which font/mapping was used. For example, if a dedicated font is used for the page numbers, _{nb}_ would become _\x01\x02\x03\x04_. The same string is created if another text starts with four different characters (which is likely), which would be replaced as well then. The simplest solution to prevent conflicts is that all characters of the `str_alias_nb_pages` and its resulting digits always get the same mapping.

To avoid having to embed all 125 characters, `font["subset"]` should be of another type. I guess `dict.get()` is also faster than `list.index()` 😉 I'll give it a try.

Thanks for your feedback!

Cheers,
Moe

Lucas-C commented on fpdf/ttfonts.py:
```diff
@@ -414,7 +418,7 @@ def extractInfo(self):
         self.getHMTX(numberOfHMetrics, numGlyphs, glyphToChar, scale)
 
-    def makeSubset(self, file, subset):
+    def makeSubset(self, file, subset, remap = {}):
```
Could you maybe add a short docstring explaining what is the role of this `remap` dict,
and what are its keys & values?

Also, please do not set a mutable default value as it can be very error prone:
https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

Lucas-C reviewed (changes requested):
Wow, this looks really promising!
Thank you for tackling this subject 🙏

Could you please add some tests that produce PDF files,
to get an idea of what users will be able to produce?
_cf._ https://pyfpdf.github.io/fpdf2/Development.html#testing

Lucas-C commented:
@all-contributors please add @moe-25 for research!

Lucas-C commented:
@all-contributors please add @moe-25 for research !

Lucas-C commented:
@all-contributors please add @moe-25 for research

alexp1917 reviewed (approved):
given that this can be formatted and made to pass tests, looks great. while i would like for there to be many more comments on clarify as to what is going on, i dont get it now, and if i dont after this PR, this changes nothing for me.

part of the mystery is the problem is the (nonexistant) code structure. so if the explanation of the code actually comes with a suggestion on how to re-organize it too, thats fine by me. thanks for the contribution! neat stuff

alexp1917 commented:
and yes, tests are important, but tests exist to test some detail in the implementation, which as far as I'm concerned, is a pretty untestable black box.

moe-25 commented:
I updated the branch to include the discussed aspects.

Most notably, `font['subset']` is no longer a list but instances a new class (that's why I noticed it shouldn't be pickled in #187). I'm pretty sure that `SubsetMap` can be optimized but it shouldn't make things worse (earlier, it stored a copy of the entire text because every character was appended regardless of its previous existence). I also added some demos/tests to print some DejaVuSans glyphs with unicodes higher 0xFFFF.

Finally, we might discuss whether to change `TTFontFile.subset`'s signature. My first though was just adding a new parameters, but now I think changing `subset` to actually be the mapping dictionary is more suitable. The current version auto-detects the given type.

Happy to hear your opinion again. Awesome library!

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -121,6 +121,43 @@ class ToCPlaceholder(NamedTuple):
+
+        # sort list to ease deletion once _next
+        # becomes higher than first reservation
+        self._reserved = sorted(identities)
```
This really is an edge case, but `SubsetMap.pick` will no behave as expected if `identities` contain duplicates here, _e.g._ if `identities == (0, 0, 0, 2)`

moe-25 replied:
Indeed, I have not considered this case. The same applies to negative values. But apart from the fact that `_reserved` would not be cleared properly, it should actually have no noticeable impact. I will now point this out in the docstring and weakly type `identities` as `Set[int]`. `sbarr` will only hold acceptable values in practice anyway. Do you think this is enough or should it still be checked more aggressively?

Lucas-C replied:
That's perfect for me, no need for more defensive programming here I think 😉

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -2652,6 +2710,24 @@ def _putfonts(self):
+                for code in subset:
+                    code_mapped = subset.get(code)
+                    if code > 0xFFFF:
+                        # Calculate suggorate pair
```
Minor typo: "surrogate"

Lucas-C commented on test/fonts/test_font_remap.py:
```diff
@@ -0,0 +1,83 @@
+from pathlib import Path
```
Really great tests, good job! 👍

moe-25 commented:
The mentioned change in `SubsetMap` and a bugfix in the surrogate calculation are still pending.

I also found a bug in `text()` that obviously didn't come into play before. Since a separate PR is not worthwhile, I would fix that right away.

Demo against master:
```
pdf = FPDF()
pdf.set_compression(False)
pdf.add_font('DejaVuSans', '', 'DejaVuSans.ttf', uni=True)
pdf.set_font('DejaVuSans')
pdf.add_page()
pdf.text(50, 50, 'Te()st') # Parentheses won't be escaped correctly
pdf.output('Demo.pdf')
```

This feature now seems to be complete from my point of view. If you agree, I'd be happy to rebase everything.

Cheers,
Moe

Lucas-C commented:
Merged!

Thanks a lot @moe-25 😉

moe-25 commented:
> The mentioned change in `SubsetMap` and a bugfix in the surrogate calculation are still pending.

No wait... :fearful: There were still bugs I wanted to fix today. To be honest, I don't know if I can just push them into my branch now or if it requires a new PR. Can you help me?

Lucas-C commented:
Wooops, sorry, my mistake!

I thought this was ready to be merged 🤭

No worries though, I haven't made a new release yet.

It would be best to open a new PR now, but I think you can rebase your branch and open a new PR based on it.
